students/timm67/session07/html_render.py

```diff
@@ -7,25 +7,201 @@
 
 # This is the framework for the base class
 class Element(object):
-    object.content = []
+    _doctype = '<!DOCTYPE html>'
+    _tag = 'html'
+    indent = '    '
+    _attributes = {}
 
-    def __init__(self, content=None):
-        if content is not None:
-            self.append(content)
+    def __init__(self, content_in=None, **kwargs):
+        if content_in is not None:
+            self._content = [content_in]
+        else:
+            self._content = []
+
+        # save attributes passed in as kwargs
+        if len(kwargs.items()) > 0:
+            for key, value in kwargs.items():
+                self._attributes[key] = value
+        else:
+            self._attributes = {}
 
     def append(self, new_content):
-        self.content.append(new_content)
+        if new_content is not None:
+            self._content.append(new_content)
+
+    def render_attributes(self, out_fd, indent_in):
+        num_attrib = len(self._attributes)
+        if(num_attrib > 0):
+            out_fd.write('{0}<{1} '.format(indent_in, self._tag))
+            for key, value in self._attributes.items():
+                out_fd.write('{0}="{1}" '.format(key, value))
+        else:
+            out_fd.write('{0}<{1}>'.format(indent_in, self._tag))
+        return num_attrib
 
-    def render(self, out_file):
+    def render(self, out_fd, indent_in=''):
         try:
-            # TODO: may want to append to already opened file (i.e. call render multiple times)
-            with open(out_file, 'w') as fd:
-                for line in self.content:
-                    try:
-                        fd.write(line)
-                    except IOError:
-                        print("I/O Error with file [{0}] on writeline".format(out_file))
-                        return
+            # render DOCTYPE
+            if isinstance(self, Html) is True:
+                out_fd.write('{0}\n'.format(self._doctype))
+
+            num_attr = self.render_attributes(out_fd, indent_in)
+            if (num_attr > 0):
+                out_fd.write('>\n')
+            else:
+                out_fd.write('\n')
```

Comment:
The `render_attributes()` is definitely a good change, but I'm not sure why you put these lines (49-62) here, outside `render_attributes()`, when they seem to fit in the `render_attributes()` just fine...?

Reply:
Minor change I'd like to forgo to avoid code churn. Yes could be in render_attributes()